dropshot_endpoint/src/endpoint.rs

```diff
@@ -479,6 +491,204 @@ impl<'ast> EndpointParams<'ast> {
     }
 }
 
+/// Perform syntactic validation for an argument or return type.
+///
+/// This returns the input type if it is valid.
+fn validate_param_ty<'ast>(
+    ty: &'ast syn::Type,
+    kind: ParamTyKind,
+    name_str: &str,
+    errors: &ErrorSink<'_, Error>,
+) -> Option<&'ast syn::Type> {
+    // Types can be arbitrarily nested, so to keep these checks simple we use
+    // the visitor pattern.
+
+    let errors = errors.new();
+
+    // This just needs a second 'store lifetime because the one inside ErrorSink
+    // is invariant. Everything else is covariant and can share lifetimes.
+    struct Visitor<'store, 'ast> {
+        kind: ParamTyKind,
+        name_str: &'ast str,
+        errors: &'ast ErrorSink<'store, Error>,
+    }
+
+    impl<'store, 'ast> Visit<'ast> for Visitor<'store, 'ast> {
+        fn visit_bound_lifetimes(&mut self, i: &'ast syn::BoundLifetimes) {
+            let name_str = self.name_str;
+            let kind = self.kind;
+            self.errors.push(Error::new_spanned(
+                i,
+                format!(
+                    "endpoint `{name_str}` must not have lifetime bounds \
+                     in {kind}",
+                ),
+            ));
+        }
+
+        fn visit_lifetime(&mut self, i: &'ast syn::Lifetime) {
+            let name_str = self.name_str;
+            let kind = self.kind;
+            if i.ident != "static" {
+                self.errors.push(Error::new_spanned(
+                    i,
+                    format!(
+                        "endpoint `{name_str}` must not have lifetime parameters \
+                         in {kind}",
+                    ),
+                ));
+            }
+        }
+
+        fn visit_ident(&mut self, i: &'ast syn::Ident) {
+            if i == "Self" {
+                let name_str = self.name_str;
+                let kind = self.kind;
+                self.errors.push(Error::new_spanned(
+                    i,
+                    format!(
+                        "endpoint `{name_str}` must not have `Self` in {kind}",
+                    ),
+                ));
+            }
+        }
+
+        fn visit_type_impl_trait(&mut self, i: &'ast syn::TypeImplTrait) {
+            let name_str = self.name_str;
+            let kind = self.kind;
+            self.errors.push(Error::new_spanned(
+                i,
+                format!(
+                    "endpoint `{name_str}` must not have impl Trait in {kind}",
+                ),
+            ));
+        }
+    }
+
+    let mut visitor = Visitor { kind, name_str, errors: &errors };
+    visitor.visit_type(ty);
+
+    // Don't return the type if there were errors.
+    (!errors.has_errors()).then(|| ty)
+}
+
+/// A representation of the RequestContext type.
+#[derive(Clone, Eq, PartialEq)]
+enum RqctxTy<'ast> {
```

Comment:
this is what we're going to add to for the trait-based approach?

Reply:
> this is what we're going to add to for the trait-based approach?

Yeah.